functions/src/glean.ts

```diff
@@ -0,0 +1,228 @@
+import Glean from "@mozilla/glean/node";
+import PingEncryptionPlugin from "@mozilla/glean/plugins/encryption";
+import axios from "axios";
+import { Mutex, Semaphore, withTimeout } from "async-mutex";
+
+import * as rallyMetrics from "./generated/rally.js";
+import * as userMetrics from "./generated/user.js";
+import * as enrollmentMetrics from "./generated/enrollment.js";
+import * as unenrollmentMetrics from "./generated/unenrollment.js";
+import * as rallyPings from "./generated/pings.js";
+
+import {
+  Uploader,
+  UploadResult,
+  UploadResultStatus,
+} from "@mozilla/glean/uploader";
+
+const GLEAN_DEBUG_VIEW_TAG = "MozillaRally";
+const GLEAN_RALLY_APP_ID = process.env.NODE_ENV !== "test" ? "rally-core" : "test-app-id";
+const GLEAN_APP_DISPLAY_VERSION = "TODO_app_version";
+const GLEAN_ENCRYPTION_JWK = {
+  crv: "P-256",
+  kid: "rally-core",
+  kty: "EC",
+  x: "m7Gi2YD8DgPg3zxora5iwf0DFL0JFIhjoD2BRLpg7kI",
+  y: "zo35XIQME7Ct01uHK_LrMi5pZCuYDMhv8MUsSu7Eq08",
+};
+
+const GLEAN_DEFAULT_TIMEOUT = 10000;
+const gleanLock = withTimeout(new Mutex(), GLEAN_DEFAULT_TIMEOUT); // Lock on global Glean instance, metrics, and pings
+const submitPingFlag = withTimeout(new Semaphore(1), GLEAN_DEFAULT_TIMEOUT); // Allow Glean to signal once ping is sent
```

Comment:
Do we still need these now that we're not re-initializing Glean?

Reply:
@rhelmer Now that we're not re-initializing Glean, we are instead sharing one single Glean instance across potentially multiple concurrent cloud function invocations. The mutex ensures that concurrent invocations don't step on each other's toes.

Also, since we can't `await Glean.shutdown()`, we need another way to signal to the "main thread" that the ping has been sent and it can exit. The semaphore acts as a signal: at the end of the CustomUploader it unblocks the main pinging function and allows it to return.

Reply:
Thanks!